File: leetcode/design/281_Zigzag_Iterator/codes/design.py
class ZigzagIterator:
    def __init__(self, v1: list[int], v2: list[int]):
        self.data = [v1, v2]
        self.ptrs = [0, 0]
        self.turn = 0

    def next(self) -> int:
        for _ in range(2):
            i = self.turn % 2
            if self.ptrs[i] < len(self.data[i]):
                res = self.data[i][self.ptrs[i]]
                self.ptrs[i] += 1
                self.turn += 1
                return res
            self.turn += 1
        return -1

    def hasNext(self) -> bool:
        return any(self.ptrs[i] < len(self.data[i]) for i in [0, 1])


# Your ZigzagIterator object will be instantiated and called as such:
# i, v = ZigzagIterator(v1, v2), []
# while i.hasNext(): v.append(i.next())

File: leetcode/design/281_Zigzag_Iterator/codes/design.cpp
class ZigzagIterator {
    vector<int>& v1;
    vector<int>& v2;
    int i1 = 0, i2 = 0;
    int turn = 0;

public:
    ZigzagIterator(vector<int>& v1, vector<int>& v2) : v1(v1), v2(v2) {}

    int next() {
        for (int _ = 0; _ < 2; ++_) {
            if (turn % 2 == 0 && i1 < v1.size()) {
                turn++;
                return v1[i1++];
            } else if (turn % 2 == 1 && i2 < v2.size()) {
                turn++;
                return v2[i2++];
            }
            turn++;
        }
        return -1;
    }

    bool hasNext() {
        return i1 < v1.size() || i2 < v2.size();
    }
};


/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator i(v1, v2);
 * while (i.hasNext()) cout << i.next();
 */